benchmark/mnist_benchmark.py

```diff
@@ -0,0 +1,660 @@
+                            "coreset_size_1": {
+                                "run_1": accuracy_value,
+                                "run_2": accuracy_value,
+                                ...
+                            },
+                            ...
+                        },
+                        ...
+                    }
+                    Each algorithm contains coreset sizes as keys, with values being
+                    dictionaries of accuracy results from different runs.
+    """
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    file_name = "mnist_benchmark_results.json"
+    with open(os.path.join(base_dir, file_name), "w", encoding="utf-8") as f:
+        json.dump(results, f, indent=2)
+
+    print(f"Data has been saved to {file_name}")
+
+
+# pylint: disable=too-many-locals
+def main() -> None:
+    """
+    Perform the benchmark for multiple solvers, coreset sizes, and random seeds.
+
+    The function follows these steps:
+    1. Prepare and load the MNIST datasets (training and test).
+    2. Perform dimensionality reduction on the training data using PCA.
+    3. Initialise solvers for data reduction.
+    4. For each solver and coreset size, reduce the dataset and train the model
+       on the reduced set.
+    5. Train the model and evaluate its performance on the test set.
+    6. Save the results, which include test accuracy for each solver and coreset size.
+    """
+    (train_data_jax, train_targets_jax, test_data_jax, test_targets_jax) = (
+        prepare_datasets()
+    )
+    train_data_pca = Data(pca(train_data_jax))
+
+    results = {}
+
+    config = {
+        "epochs": 100,
+        "batch_size": 8,
+        "learning_rate": 1e-3,
+        "weight_decay": 1e-5,
+        "patience": 5,
+        "min_delta": 0.001,
+    }
+
+    # Run the experiment with 5 different random keys
+    for i in range(5):
+        print(f"Run {i + 1} of 5:")
+        key = jax.random.PRNGKey(i)
+        solvers = initialise_solvers(train_data_pca, key)
+        for getter in solvers:
+            for size in [25, 50, 100, 500, 1000, 5000]:
+                name, solver = getter(size)
+                start_time = time.perf_counter()
+                coreset, _ = solver.reduce(train_data_pca)
+
+                coreset_indices = coreset.nodes.data
+
+                train_data_coreset = train_data_jax[coreset_indices]
+                train_targets_coreset = train_targets_jax[coreset_indices]
+
+                # Adjust batch size based on size
+                config["batch_size"] = min(len(coreset_indices) // 2, 64)
+
+                data_bundle = {
+                    "data": train_data_coreset,
+                    "targets": train_targets_coreset,
+                    "test_data": test_data_jax,
+                    "test_targets": test_targets_jax,
+                }
+                result = train_model(data_bundle, key, config)
+                if name not in results:
```

Comment:
Can you add a comment above this if statement describing its purpose? I think it's creating the dictionaries if the solver name does not exist yet in `results`, but it's not entirely clear. 

Same for `if size not in results[name]:` below